### PR TITLE
Fix #5838: Run batch operations at the end of config import.

### LINKED ISSUE #5838: Error when modules get installed via config:import after PR #5714  (locale and language config collections)
**Describe the bug**
After updating to drush 12, and PR https://github.com/drush-ops/drush/pull/5714 included in drush 12.1.3 specifically, it looks like there can be a side effect with language config collections in some situations.

I am getting this error on a multilingual site during `config:import`:
> [warning] Trying to access array offset on value of type null ConfigImporter.php:353
> > TypeError: array_diff(): Argument #2 must be of type array, null given in web/core/lib/Drupal/Core/Config/ConfigImporter.php on line 353 #0 web/core/lib/Drupal/Core/Config/ConfigImporter.php(353): array_diff()
See full log below.

At the beginning of `config:import`, `locale_system_update` gets called if modules need to be enabled.
This can cause some _new_ config collections to be added (during the batch process that https://github.com/drush-ops/drush/pull/5714 added) and if so, leads to the error above.
So initially, we need a language but without any corresponding config collection records (in the `config` table).

This may not be an issue with drush itself but drupal core.

**To Reproduce**
In my case I have an `mk` language in drupal and _no_ corresponding `language.mk` config collection (in the `config` table) initially. Then when running `config:import`, devel gets enabled which imports config translations.
See full log below.

This may not be easy to reproduce from a clean setup... but I'll try to do it when I have some time.

**Expected behavior**
No error

**Actual behavior**
An error

**Workaround**
I noticed calling `locale_config_batch_update_components` before `config:import` runs (in a post-update hook for example) prevents the issue to happen because it will create several records with the language.mk collection in config table (but it is not guaranteed to do so).

### System Configuration
| Q               | A
| --------------- | ---
| Drush version?  | >= 12.1.3 (with PR https://github.com/drush-ops/drush/pull/5714 included)
| Drupal version? | 10.1.7
| PHP version     | 8.1
| OS?             | Linux

**Additional information**
Here is the relevant output from config:import:

```
// Import the listed configuration changes?: yes.                              

>  [notice] Synchronized extensions: uninstall allowed_formats.
>  [notice] Synchronized extensions: uninstall f_pt_cta.
>  [notice] Synchronized extensions: install stage_file_proxy.
>  [info] Executing: /vendor/bin/drush batch-process 145 --uri=default
> > 
> >  [notice] Message: No configuration objects have been updated.
> > 
>  [notice] Synchronized extensions: install f_dev.
>  [info] Executing: /vendor/bin/drush batch-process 146 --uri=default
> > 
> >  [notice] Message: No configuration objects have been updated.
> > 
>  [notice] Synchronized extensions: install devel.
>  [info] Executing: /vendor/bin/drush batch-process 147 --uri=default
> > 
> >  [notice] Message: No configuration objects have been updated.
> > 
>  [warning] Undefined array key "language.mk" ConfigImporter.php:353
>  [warning] Trying to access array offset on value of type null ConfigImporter.php:353
> TypeError: array_diff(): Argument #2 must be of type array, null given in /web/core/lib/Drupal/Core/Config/ConfigImporter.php on line 353 #0 /web/core/lib/Drupal/Core/Config/ConfigImporter.php(353): array_diff()
> #1 /web/core/lib/Drupal/Core/Config/ConfigImporter.php(650): Drupal\Core\Config\ConfigImporter->getUnprocessedConfiguration()
> #2 /web/core/lib/Drupal/Core/Config/ConfigImporter.php(561): Drupal\Core\Config\ConfigImporter->processConfigurations()
> #3 /vendor/drush/drush/src/Commands/config/ConfigImportCommands.php(259): Drupal\Core\Config\ConfigImporter->doSyncStep()
> #4 /vendor/drush/drush/includes/drush.inc(62): Drush\Commands\config\ConfigImportCommands->doImport()
> #5 /vendor/drush/drush/includes/drush.inc(53): drush_call_user_func_array()
> #6 /vendor/drush/drush/src/Commands/config/ConfigImportCommands.php(230): drush_op()
> #7 [internal function]: Drush\Commands\config\ConfigImportCommands->import()
> #8 /vendor/consolidation/annotated-command/src/CommandProcessor.php(276): call_user_func_array()
> #9 /vendor/consolidation/annotated-command/src/CommandProcessor.php(212): Consolidation\AnnotatedCommand\CommandProcessor->runCommandCallback()
> #10 /vendor/consolidation/annotated-command/src/CommandProcessor.php(176): Consolidation\AnnotatedCommand\CommandProcessor->validateRunAndAlter()
> #11 /vendor/consolidation/annotated-command/src/AnnotatedCommand.php(391): Consolidation\AnnotatedCommand\CommandProcessor->process()
> #12 /vendor/symfony/console/Command/Command.php(326): Consolidation\AnnotatedCommand\AnnotatedCommand->execute()
> #13 /vendor/symfony/console/Application.php(1081): Symfony\Component\Console\Command\Command->run()
> #14 /vendor/symfony/console/Application.php(320): Symfony\Component\Console\Application->doRunCommand()
> #15 /vendor/symfony/console/Application.php(174): Symfony\Component\Console\Application->doRun()
> #16 /vendor/drush/drush/src/Runtime/Runtime.php(110): Symfony\Component\Console\Application->run()
> #17 /vendor/drush/drush/src/Runtime/Runtime.php(40): Drush\Runtime\Runtime->doRun()
> #18 /vendor/drush/drush/drush.php(139): Drush\Runtime\Runtime->run()
> #19 /vendor/drush/drush/drush(4): require('...')
> #20 /vendor/bin/drush(119): include('...')
> #21 {main}
TypeError: array_diff(): Argument #2 must be of type array, null given in array_diff() (line 353 of /web/core/lib/Drupal/Core/Config/ConfigImporter.php).
>  [warning] Drush command terminated abnormally.

In SiteProcess.php line 214:
                                                                               
  [Symfony\Component\Process\Exception\ProcessFailedException]                 
  The command "/vendor/bin/drush config:import --verbose --yes -  
  -uri=default" failed.                                                        
                                                                               
  Exit Code: 1(General error)

```

### PR DESCRIPTION
**Pull request by vever001**
See https://github.com/drush-ops/drush/issues/5838

### CODE REVIEW

File: src/Commands/config/ConfigImportCommands.php [hunk 1]
```diff
@@ -260,22 +260,30 @@ public function doImport($storage_comparer): void
...
                     }
                     // Clear the cache of the active config storage.
                     $this->getConfigCache()->deleteAll();
+
+                    // Importing config might trigger batch operations (such as when installing and uninstalling modules).
+                    // Let's process them here.
+                    // @see \Drush\Commands\pm\PmCommands::install()
+                    if (batch_get()) {
+                        drush_backend_batch_process();
+                    }
```
Comment (vever001) [inline src/Commands/config/ConfigImportCommands.php]: Whoops, didn't mean to add it twice, I'll amend sry.

File: src/Commands/config/ConfigImportCommands.php [hunk 2]
```diff
@@ -260,22 +260,23 @@ public function doImport($storage_comparer): void
...
-                    $this->logger()->success('The configuration was imported successfully.');
+                }
+
+                $this->logger()->success('The configuration was imported successfully.');
+
+                // Importing config might trigger batch operations (such as when installing and uninstalling modules).
+                // @see \Drush\Commands\pm\PmCommands::install()
+                if (batch_get()) {
+                    $this->logger()->info('Running batch operations...');
+                    drush_backend_batch_process();
```
Comment (vever001) [inline src/Commands/config/ConfigImportCommands.php]: Thinking aloud: should we move this to a `finally` block below?

Comment (vever001) [reply src/Commands/config/ConfigImportCommands.php]: I moved it.
Not 100% sure but can a `ConfigException` happen in a middle of config import, on a specific file?

**Comment by claudiu-cristea** [comment]
As I've mentioned in https://github.com/drush-ops/drush/issues/5838#issuecomment-1959848475, I've manually tested this with the same code that required #5714. Everything works as expected. I see now the batch ops running at the end of config import

However, there are still 2 issues:

* https://github.com/drush-ops/drush/pull/5869/files#diff-e4abd35e5f310a7c79c50fe7bfd4e5a624b1520464853849d77884be1843b2d5R289 this notice is not shown
* We need to fix the `::testBatchImportTranslationsOnInstall()` test that was recently re-enabled

**Comment by vever001** [comment]
Hi @claudiu-cristea 
1. I tested the patch against the branch that you mentioned.
It appears that search_api runs batch operations [here](https://git.drupalcode.org/project/search_api/-/blob/8.x-1.x/src/Task/TaskManager.php?ref_type=heads#L316) when a search index gets updated.
Which explains why batch operations run during sync steps on that branch, and not at the end, and also why we don't see the notice.
Aside from this particular case, the message is displayed and batch operations happen at the very end.
2. I was not able to reproduce the error shown in the CI, can we re-run it maybe?
But, I've see this error on a colleague's machine so maybe some bug is hiding there.

**Comment by claudiu-cristea** [comment]
@vever001 

> I tested the patch against the branch that you mentioned.
It appears that search_api runs batch operations [here](https://git.drupalcode.org/project/search_api/-/blob/8.x-1.x/src/Task/TaskManager.php?ref_type=heads#L316) when a search index gets updated.
Which explains why batch operations run during sync steps on that branch, and not at the end, and also why we don't see the notice.

Fair enough

>  was not able to reproduce the error shown in the CI, can we re-run it maybe? But, I've see this error on a colleague's machine so maybe some bug is hiding there.

It seem a PHP 8.3 error. Maybe you should try with that PHP version

**Comment by weitzman** [comment]
I finally got Xdebug disabled and now this is green. Drupal has a bug with Xdebug, but not Drush so in this goes. Thanks all.

**Pull request merged**